detect.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"path/filepath"
+
+	"github.com/cloudfoundry/packit"
+)
+
+//go:generate faux --interface VersionParser --output fakes/version_parser.go
+type VersionParser interface {
+	ParseVersion(path string) (version string, err error)
+}
+
+type BuildPlanMetadata struct {
+	VersionSource string `toml:"version-source"`
+	Build         bool   `toml:"build"`
+}
+
+func Detect(buildpackYAMLParser VersionParser) packit.DetectFunc {
+	return func(context packit.DetectContext) (packit.DetectResult, error) {
+		var requirements []packit.BuildPlanRequirement
+		version, err := buildpackYAMLParser.ParseVersion(filepath.Join(context.WorkingDir, "buildpack.yml"))
+		if err != nil {
+			return packit.DetectResult{}, err
+		}
+
+		if version != "" {
+			requirements = append(requirements, packit.BuildPlanRequirement{
+				Name:    GoDependency,
+				Version: version,
+				Metadata: BuildPlanMetadata{
+					VersionSource: "buildpack.yml",
+					Build:         true,
+				},
+			})
+		}
+
+		return packit.DetectResult{
```

Comment:
This is a departure from the way detect previously worked and breaks the `go-mod` integration tests.
This buildpack used to always require itself as a `build` dependency as seen [here](https://github.com/paketo-buildpacks/go-compiler/blob/8fa99f2a005a609d4599972598f4b285671cbb53/cmd/detect/main.go#L53).

The behavior implemented in this re-write seems preferable to the old behavior, but we should update the go-mod buildpack to require `go` as a build time dependency first [here](https://github.com/paketo-buildpacks/go-mod/blob/26260ef0b41674eca45c3ea61ffd4abe16109ed1/cmd/detect/main.go#L51)

Reply:
I've added https://github.com/paketo-buildpacks/go-mod/pull/36 and https://github.com/paketo-buildpacks/dep/pull/25 to resolve this.